### PR TITLE
Merge 3.1

### PR DESCRIPTION
**Pull request by wallyworld**
Merge 3.1 - no conflicts

https://github.com/juju/juju/pull/15243
https://github.com/juju/juju/pull/15242
https://github.com/juju/juju/pull/15219

### CODE REVIEW

File: upgrades/steps_311.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2022 Canonical Ltd.
```
Comment (anvial) [inline upgrades/steps_311.go:1]: Maybe 2023?

**Review by anvial — APPROVED** [review]
Built, checked bootstrap (lxd, aws) and deploy.

LGTM

PS: One minor issue about comments

File: upgrades/steps_311.go (hunk 1 above)
Comment (wallyworld) [reply upgrades/steps_311.go:1]: actually, i needed to delete this and the other upgrade steps

**Pull request merged**